tl.tl

```diff
@@ -458,6 +460,9 @@ do
 
       local function end_token_prev(kind: TokenKind)
          local tk = input:sub(ti, i - 1)
+         if kind == "identifier" and keywords[tk] then
```

Comment:
This check shouldn't be added here. The reason why I split `end_token` into multiple tokens was precisely to specialize their behavior and avoid this extra test in all non-`identifier` states.

I suspect the reason you added this here is because of the test for the very last token that happens near the end of `lex` which is assuming `end_token_prev`. Instead of adding a test here that would add a cost for every token, that final test should instead check which of `end_token_prev` or `end_token_identifier` should be called, based on the current `state`.